date.go

```diff
@@ -30,21 +35,33 @@ type DateRule struct {
 // An empty value is considered valid. Use the Required rule to make sure a value is not empty.
 func Date(layout string) DateRule {
 	return DateRule{
-		layout:       layout,
-		message:      "must be a valid date",
-		rangeMessage: "the data is out of range",
+		layout:   layout,
+		err:      ErrDateInvalid,
+		rangeErr: ErrDateOutOfRange,
 	}
 }
 
 // Error sets the error message that is used when the value being validated is not a valid date.
 func (r DateRule) Error(message string) DateRule {
-	r.message = message
+	r.err.SetMessage(message)
+	return r
+}
+
+// ErrorObject sets the error struct that is used when the value being validated is not a valid date..
+func (r DateRule) ErrorObject(err Error) DateRule {
```

Comment:
Yea, I agree with you, but if I change it, in all other methods of rules that set error properties we must assert that if it is `Error` then set its value, otherwise return an error. everything can be complicated unless we change the `Error` type to be an interface with our needed methods.

Reply:
Yes, we need an interface here:
```go
type Error interface {
    Error() string
    Code() string
    Message() string
    SetMessage(string)
    Params() map[string]interface{}
    SetParams(map[string]interface{})
}

type ErrorObject struct {...} // the current Error struct is renamed as ErrorObject or something better
```

Reply:
@qiangxue I think Error interface should be something like this:
```go
type Error interface {
    Error() string
    Code() string
    SetCode(string) Error
    Message() string
    SetMessage(string) Error
    Params() map[string]interface{}
    SetParams(map[string]interface{}) Error
    AddParam(string,string) Error
}

```

this is because if we do not return Error and get a pointer in seters, so in `NewError` we have to return a pointer to the `ErrorObject`, so all predefined errors will be pointers.

What do you think?

Reply:
I'd like to keep the interface minimal. That's why `SetCode` and `AddParams` are not in my design, because they are not used/required anywhere.
Yes, `NewError` should return a pointer.

Reply:
But if return a pointer, changes to predefined errors affect base error, e.g 
```go
myErr=validation.ErrEmail
err.SetParams(map[string]interface{}{"min": min, "max": max})
```
this line set params on-base error also. we should get a copy of it.

Reply:
You're right. We should add a `Clone()` method to `ErrorObject` (not to the `Error` interface).

Reply:
To avoid frequent memory allocations, we can call `Clone()` only when we are trying to update the predefined errors. For example, `rule.err.Clone().SetMessage(...)`

Reply:
We are assigning the `Error` interface to each validation rule's `err` property, so on each validation rule creation, we must first check if our predefined error object is cloneable (since our users can change predefined errors and do not implement Clone method) then clone it, otherwise, simply assign it. our rule initialization is finally something like this:
```go
ErrDateInvalid = NewError("validation_date_invalid", "must be a valid date")

func Date(layout string) DateRule {
	return DateRule{
		layout:   layout,
		err:      TryToClone(ErrDateInvalid),
		rangeErr: TryToClone(ErrDateOutOfRange),
	}
}

type CloneableErrorObject interface {
	Clone() Error
}

func TryToClone(err Error) Error {
	if cloneable, ok := err.(CloneableErrorObject); ok {
		return cloneable.Clone()
	}

	return err
}
``` 
I think we should remove all of these complexities and just change the `ErrorObject` setters signature to get just copy, not a pointer, e.g:

```go
// SetCode set the error's translation code.
func (e ErrorObject) SetCode(code string) Error {
	e.code = code
        return e
}
```

Reply:
You're right. Let's modify the signature `SetCode` and `SetParams` so that they return `Error`. Thanks!

Reply:
yep, you're welcome :)